lib/mutgen.c

```diff
@@ -433,28 +710,289 @@ mutgen_generate(mutgen_t *self, tsk_table_collection_t *tables, int flags)
+    } else {
+        ai = mutation_model_allele_index(self->model,
+                site->ancestral_state, site->ancestral_state_length);
+    }
+
+    /* Create a mapping from mutations to nodes in bottom_mutation. If we see
+     * more than one mutation at a node, the previously seen one must be the
+     * parent of the current since we assume they are in order. */
+    j = 0;
+    for (mut = site->mutations; mut != NULL; mut = mut->next) {
+        mut->id = mutation_id + j;
+        u = mut->node;
+        assert((tsk_size_t) u < num_nodes);
+        while (u != TSK_NULL && bottom_mutation[u] == NULL) {
+            u = parent[u];
+        }
+        } else {
+            parent_mut = bottom_mutation[u];
+            mut->parent = parent_mut->id;
+            if (mut->parent >= mut->id || mut->time > parent_mut->time ||
+                    (parent_mut->new && !mut->new)) {
+                ret = MSP_ERR_MUTATION_GENERATION_OUT_OF_ORDER;
+                goto out;
+            }
+            if (mut->new) {
+                pa = mutation_model_allele_index(self->model,
+                        parent_mut->derived_state, parent_mut->derived_state_length);
+            }
+        }
+        keep_mutation = true;
+        if (mut->new) {
+            if (pa < 0) {
+                /* only error if we are actually trying to mutate an unknown allele */
+                ret = MSP_ERR_UNKNOWN_ALLELE;
+                goto out;
+            }
+            da = mutgen_transition_allele(self, pa);
+            if (da == pa) {
+                // remove mut from the chain
+                delete_mutation(site, mut);
+                keep_mutation = false;
+            } else {
+                memcpy(mut->derived_state, model->alphabet + da, sizeof(*mut->derived_state));
+            }
+        }
+        if (keep_mutation) {
+            bottom_mutation[mut->node] = mut;
+            j++;
+        }
+    }
+    /* Reset the mapping for the next site */
+    for (mut = site->mutations; mut != NULL; mut = mut->next) {
+        bottom_mutation[mut->node] = NULL;
+    }
+
+out:
+    return ret;
+}
+
+static int MSP_WARN_UNUSED
+mutgen_apply_mutations(mutgen_t *self, tsk_table_collection_t *tables)
+{
+    int ret = 0;
+    const tsk_id_t *I, *O;
+    const tsk_edge_table_t edges = tables->edges;
+    const tsk_node_table_t nodes = tables->nodes;
+    const tsk_id_t M = (tsk_id_t) edges.num_rows;
+    tsk_id_t tj, tk, mutation_id;
+    tsk_id_t *parent = NULL;
+    mutation_t **bottom_mutation = NULL;
+    double left, right;
+    avl_node_t *avl_node;
+    site_t *site;
+    size_t j;
+
+    parent = malloc(nodes.num_rows * sizeof(*parent));
+    bottom_mutation = malloc(nodes.num_rows * sizeof(*bottom_mutation));
+    if (parent == NULL || bottom_mutation == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(parent, 0xff, nodes.num_rows * sizeof(*parent));
```

Comment:
``memset(bottom_mutation, 0, nodes.num_rows * sizeof(*parent));`` will do the trick here, as  0 == NULL.